### PR TITLE
Optimizer num_step_per_communication behavior change and test

### PR DESCRIPTION
**Pull request by hanbinhu**

### CODE REVIEW

File: test/torch_optimizer_test.py [hunk 1]
```diff
@@ -37,10 +37,49 @@ class LinearNet(nn.Module):
...
+        self._num_parameters = input_dim*output_dim
 
     def forward(self, x):
         return self.fc(x)
 
+    @property
+    def num_parameters(self):
+        return self._num_parameters
+
+# A deep linear model for testing
```
Comment (Bluefog-Lib) [inline test/torch_optimizer_test.py:49]: not deep? just three layers.

File: test/torch_optimizer_test.py [hunk 2]
```diff
@@ -37,10 +37,49 @@ class LinearNet(nn.Module):
...
+    
+    @property
+    def num_parameters(self):
+        return self._num_parameters
+
+class HierarchicalLinearNet(nn.Module):
+    def __init__(self, input_dim, output_dim):
+        super(HierarchicalLinearNet, self).__init__()
+        self.fc1 = LinearNet(input_dim, output_dim)
+        self.fc2 = LinearNet(output_dim, output_dim)
```
Comment (Bluefog-Lib) [inline test/torch_optimizer_test.py:71]: Maybe it is better with one LinearNet and another DuplicatedLinearNet? Because LinearNet is almost same as nn.Linear.

File: test/torch_optimizer_test.py [hunk 3]
```diff
@@ -317,6 +356,80 @@ def test_standard_optimizer(device, communication_type, kwargs):
...
+                     marks=pytest.mark.skip(reason="AWC doesn't converge for hierarchical model.")))
+    hierarchical_model_scenarios.append(
+        pytest.param("GPU", bf.CommunicationType.neighbor_allreduce, {"ATC": True},
+                     id="ATC Neighbor Allreduce on GPU",
+                     marks=pytest.mark.skip(reason="ATC doesn't converge for hierarchical model.")))
+    hierarchical_model_scenarios.append(
+        pytest.param("GPU", "gradient.allreduce", {}, id="Gradient Allreduce on GPU"))
+    hierarchical_model_scenarios.append(
+        pytest.param("GPU", "win.put", {}, id="Window put on GPU",
+                     marks=pytest.mark.skip(reason="Multiple win_put optimizer tests will fail")))
```
Comment (Bluefog-Lib) [inline test/torch_optimizer_test.py:386]: nit: new line

File: test/torch_optimizer_test.py [hunk 4]
```diff
@@ -317,6 +356,80 @@ def test_standard_optimizer(device, communication_type, kwargs):
...
+@pytest.mark.parametrize("device,communication_type,kwargs", hierarchical_model_scenarios)
+def test_optimizer_for_hierarchical_model(device, communication_type, kwargs):
+    atc_style = kwargs["ATC"] if "ATC" in kwargs else False
```
Comment (Bluefog-Lib) [inline test/torch_optimizer_test.py:389]: simplify it into kwargs.get("ATC", False).  Same comments for other places.

File: test/torch_optimizer_test.py [hunk 5]
```diff
@@ -317,6 +356,80 @@ def test_standard_optimizer(device, communication_type, kwargs):
...
+    error_threshold = kwargs["error_threshold"] if "error_threshold" in kwargs else 1.5
+
+    problem_builder, train_dataloader, test_dataloader, model, optimizer, num_epochs = \
+        problem_setup(HierarchicalLinearNet)
+
+    isCUDA = device=="GPU"
```
Comment (Bluefog-Lib) [inline test/torch_optimizer_test.py:395]: This repeats many times and usage should always be the same. So better to create a global util function:

def pin_model_to_device(device):
    ....

File: bluefog/torch/optimizers.py [hunk 6]
```diff
@@ -335,42 +352,42 @@ def __init__(self, params, model, communication_type, num_steps_per_communicatio
 
     def _register_hooks(self):
         for model in self._models:
-            for parent_name, layer in _named_leaf_module(model):
-                layer.register_forward_hook(self._make_hook(parent_name))
+            model.register_forward_hook(self._make_hook())
+            for _, layer in _named_leaf_module(model):
+                #layer.register_forward_hook(self._make_hook(parent_name))
                 for _, p in layer.named_parameters():
                     self._requires_update.add(p)
```
Comment (Bluefog-Lib) [inline bluefog/torch/optimizers.py:359]: why not use model.named_parameters() to set the self. _requires_update

File: bluefog/torch/optimizers.py [hunk 7]
```diff
@@ -335,42 +352,42 @@ def __init__(self, params, model, communication_type, num_steps_per_communicatio
 
     def _register_hooks(self):
         for model in self._models:
-            for parent_name, layer in _named_leaf_module(model):
-                layer.register_forward_hook(self._make_hook(parent_name))
+            model.register_forward_hook(self._make_hook())
```
Comment (Bluefog-Lib) [inline bluefog/torch/optimizers.py:355]: Add a comment here about why model forward is sufficient

File: docs/faq.rst [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+FAQ
+===
+
+.. contents:: Question Lists
+  :local:
+
+Why does the warning related to ``num_communication_per_step`` or ``backward_passes_per_step`` pop up in Bluefog optimizer?
+--------------------------------------------------------------------------------------------
+
+During your usage of Bluefog optimizer, you may encounter the following the two types of warnings.
```
Comment (Bluefog-Lib) [inline docs/faq.rst:12]: Bluefog distributed optimizer

Comment (Bluefog-Lib) [inline docs/faq.rst:12]: the following two types (remove the middle "the")

File: docs/faq.rst [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+   counting is triggered. All other Bluefog optimizers triggers this counting
+   behavior during forward computation, while ``DistributedGradientAllreduceOptimizer`` and 
+   ``DistributedAdaptThenCombineOptimizer`` triggers during backward computation.
+   This is also reflected by the optimizer argument naming.
+   All other optimizers uses ``num_step_per_communication``, while these two optimizers uses
+   ``backward_passes_per_step``.
+   The following discussion only focuses on forward computation case, but it works for backward
+   scenario as well.
+
+Consider the following admissible code snippet.
```
Comment (Bluefog-Lib) [inline docs/faq.rst:44]: + To understand the meaning of the above two warnings, consider ..

Comment (Bluefog-Lib) [inline docs/faq.rst:44]: .. admissible code snippet for local gradient aggregation case:

File: docs/faq.rst [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+   :emphasize-lines: 3,10,13
+   
+   opt = bf.DistributedAdaptWithCombineOptimizer(
+     optimizer, model=model, communication_type=bf.CommunicationType.neighbor_allreduce,
+     num_steps_per_communication=J)
+   for _ in range(num_epochs):
+      for data, target in dataloader:
+        opt.zero_grad()
+        for i in range(J):
+          data_mini_batch = data[...]
```
Comment (Bluefog-Lib) [inline docs/faq.rst:56]: [i::J] is the valid usage for this case?

File: docs/faq.rst [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+          loss.backward()
+        opt.step() # Step Function <- Communication occurs and counting is reset.
+
+We use **F** to denote forward computation ``y=model(data_mini_batch)``,
+and use **S** to denote step function ``opt.step()`` with communication occurrence.
+For example, let's say ``J`` is 3. 
+For each batch, the calling sequence is **FFFS**, ignoring other operations.
+We can see that step function happens right after ``J=3`` forward computations.
+With that, we accumulate the update for the model parameters in all ``J=3`` mini batches locally,
+and reduce the model parameters during the step function in a AWC style.
```
Comment (Bluefog-Lib) [inline docs/faq.rst:69]: Add more sentences about if the inner iteration mismatch with the J=3. Say "FFFFS", then the error will throw at the fourth "F" because .....

File: docs/faq.rst [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+Let's check another admissible code snippet.
```
Comment (Bluefog-Lib) [inline docs/faq.rst:71]: for multi-round computation per communication case

File: docs/faq.rst [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+          opt.zero_grad()
+          loss.backward()
+          opt.step() # Step Function <- Communication occurs at Jth iteration and counting is reset.
+
+Similar as before, the calling sequence is **FsFsFS**.
+Here **s** stands for the step function without communication (reduce) occurrence.
+After ``J=3`` forward computations happen, a step function called triggers communication.
+However, all other step functions in between, denoted by **s**, doesn't trigger communication.
+With that, we update the model locally after each mini batch; and at the last mini batch,
+the model parameters are reduced with its neighbors.
```
Comment (Bluefog-Lib) [inline docs/faq.rst:95]: Same here for the wrong usage case

File: docs/faq.rst [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+These are two common usages for ``num_communication_per_step`` or ``backward_passes_per_step`` for
+Bluefog optimizer. But other usage is also allowed, as long as after ``num_communication_per_step``
```
Comment (Bluefog-Lib) [inline docs/faq.rst:98]: distributed optimizer

File: test/torch_optimizer_test.py (hunk 1 above)
Comment (hanbinhu) [reply test/torch_optimizer_test.py:49]: Done

File: test/torch_optimizer_test.py (hunk 3 above)
Comment (hanbinhu) [reply test/torch_optimizer_test.py:386]: Done

File: docs/faq.rst (hunk 8 above)
Comment (hanbinhu) [reply docs/faq.rst:12]: Done

Comment (hanbinhu) [reply docs/faq.rst:12]: Done

File: docs/faq.rst (hunk 9 above)
Comment (hanbinhu) [reply docs/faq.rst:44]: Done

Comment (hanbinhu) [reply docs/faq.rst:44]: Done.

File: docs/faq.rst (hunk 14 above)
Comment (hanbinhu) [reply docs/faq.rst:98]: Done.

File: docs/faq.rst (hunk 12 above)
Comment (hanbinhu) [reply docs/faq.rst:71]: Done.

File: test/torch_optimizer_test.py (hunk 4 above)
Comment (hanbinhu) [reply test/torch_optimizer_test.py:389]: Done.

File: test/torch_optimizer_test.py (hunk 5 above)
Comment (hanbinhu) [reply test/torch_optimizer_test.py:395]: Done.

File: bluefog/torch/optimizers.py (hunk 7 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py:355]: Done.

File: docs/faq.rst (hunk 10 above)
Comment (hanbinhu) [reply docs/faq.rst:56]: Done.

File: bluefog/torch/optimizers.py (hunk 6 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py:359]: Done.

File: test/torch_optimizer_test.py (hunk 2 above)
Comment (hanbinhu) [reply test/torch_optimizer_test.py:71]: Discussed offline. Ignore this for now.

File: docs/faq.rst (hunk 11 above)
Comment (hanbinhu) [reply docs/faq.rst:71]: Done.

File: docs/faq.rst (hunk 13 above)
Comment (hanbinhu) [reply docs/faq.rst:103]: Done.

File: bluefog/torch/optimizers.py [hunk 15]
```diff
@@ -31,6 +31,23 @@ class CommunicationType(Enum):
     allreduce = "allreduce"
     empty = "empty"
 
+_warning_message_num_step_per_communication = (
+    "Unexpected behavior:\n"
+    "  After num_steps_per_communication times of forward computation `y=model(x)` are called,\n"
+    "  an optimizer step() function must be called.\n"
+    "  It does not matter how many step() functions are called in between.\n"
+    "  Please adjust num_step_per_communication to update model parameters locally.\n"
+    "  More information can be found in the FAQ page.\n"
```
Comment (Bluefog-Lib) [inline bluefog/torch/optimizers.py:40]: Either add a TODO or give the URL link to the index page of the document website?

**Review by Bluefog-Lib — APPROVED** [review]

File: bluefog/torch/optimizers.py (hunk 15 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py:41]: Done.
